### ISSUE TITLE
incompatible_disable_deprecated_attr_params: Cleanup skylark 'attr' API

### ISSUE THREAD

**Issue by c-parsons** [bazelbuild/bazel#5818]
This is a tracking issue for offering a migration solution for
`--incompatible_disable_deprecated_attr_params`

This flag would turn down a number of deprecated APIs pertaining to attr. For example, the single_file parameter would no longer be used.

**Comment by laurentlb** [comment]
**Migration notes**

When creating attributes, for example [`attr.label_list`](https://docs.bazel.build/versions/master/skylark/lib/attr.html#label_list), use `allow_empty` instead of `non_empty`.
This can be fixed automatically by running [Buildifier](https://github.com/bazelbuild/buildtools) on the bzl files, with `--lint=fix`.

**Comment by laurentlb** [comment]
Main blocker for this is that "protobuf 3.6.1" doesn't have the fix, it still uses `single_file = True`, although it was fixed at head in August (https://github.com/protocolbuffers/protobuf/commit/d5f0dac497f833d06f92d246431f4f2f42509e04).

I'm removing the migration-ready label, because many repositories depend on protobuf.

**Comment by jin** [comment]
The latest Protobuf tag (3.6.1.3) still uses `single_file = True`. 

@acozzette could you create a new tag (3.6.1.4?) with https://github.com/protocolbuffers/protobuf/commit/d5f0dac497f833d06f92d246431f4f2f42509e04 please?

**Comment by philwo** [comment]
This flag was not flipped in time for the Bazel 0.23.0 release and will thus be postponed to Bazel 0.24.0.

**Comment by katre** [comment]
This issue was tagged as "breaking-change-0.24" but does not appear ready to be flipped in the [0.24.0 release](https://github.com/bazelbuild/bazel/issues/6968). If this is incorrect please comment on that issue and discuss with me.

**Comment by c-parsons** [comment]
This flag is ready to be flipped, I believe, but we did not actually submit a change to flip it.
My bad.
If this has missed the cut for the 0.24.0 release, then it'll have to wait for the next one. (It's not urgent)

I'll propose a commit to flip this flag.

**Comment by laurentlb** [comment]
The flag was intentionally not flipped because it would break a huge number of repositories. Now that Protobuf 3.7.0 has been released, we may be able to update them.

**Comment by jin** [comment]
3.7.0 is not buildable with Bazel yet. See https://github.com/protocolbuffers/protobuf/pull/5811

**Comment by davido** [comment]
> Now that Protobuf 3.7.0 has been released, we may be able to update them.

+1. I filed this [issue ](https://github.com/bazelbuild/bazel/issues/7611) to track `protobuf` upgrade in Bazel. I also started to work on it and realized the problem. However, given that protocolbuffers/protobuf#5811 is a trivial change and given that actually `protobuf` fork is used in Bazel tree, we could proceed with the upgrade even before this PR was merged.

I could look into upgrading `protobuf` in Bazel tree.

**Comment by laurentlb** [comment]
Many downstream projects are still broken (https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/67). We should update them to use the latest protobuf.

**Comment by davido** [comment]
> We should update them to use the latest protobuf.

Right, including upgrading Bazel itself (https://github.com/bazelbuild/bazel/pull/7681 and https://github.com/bazelbuild/bazel/pull/7775).

**Comment by laurentlb** [comment]
Status update. Flag flip still blocked by:
- [android testing](https://github.com/googlesamples/android-testing/issues/239)
- [rules_webtesting](https://github.com/bazelbuild/rules_webtesting/issues/351)

**Comment by davido** [comment]
@laurentb what about Bazel itself: pending (#7681 and #7775)? Shouldn't you mention this as well?

**Comment by laurentlb** [comment]
protobuf.bzl was updated separately (https://github.com/bazelbuild/bazel/commit/ad503849e78b98d762f03168de5a336904280150).

To identify breakages, I checked https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/114

**Comment by davido** [comment]
> protobuf.bzl was updated separately (ad50384).

Ah, right. I forgot that Bazel is using forked version of Protobuf.

**Comment by brandjon** [comment]
Looks like this breaks `rules_python` but wasn't caught by tests. Filed bazelbuild/rules_python#191.

**Comment by brandjon** [comment]
Disregard the above, user error on my part.
